README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
lme4qtl
=======

<!-- badges: start -->
[![travis-ci build status](https://travis-ci.org/variani/lme4qtl.svg?branch=master)](https://travis-ci.org/variani/lme4qtl) <!-- badges: end -->

`lme4qtl` extends the [lme4](https://github.com/lme4/lme4) R package for quantitative trait locus (qtl) mapping. It is all about the covariance structure of random effects. `lme4qtl` supports user-defined matrices for that, e.g. kinship or IBDs.

See slides [bit.ly/1UiTZvQ](http://bit.ly/1UiTZvQ) introducing the `lme4qtl` R package or read our [article](http://dx.doi.org/10.1186/s12859-018-2057-x) / [preprint](http://www.biorxiv.org/content/early/2017/08/31/139816).

<table style="width:46%;">
<colgroup>
<col width="15%" />
<col width="30%" />
</colgroup>
<thead>
<tr class="header">
<th>Package</th>
<th>Continuous response</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>stats</td>
<td><code>lm(myTrait ~ myCovariate, myData)</code></td>
</tr>
<tr class="even">
<td>lme4</td>
<td><code>lmer(myTrait ~ myCovariate + (1|myID), myData)</code></td>
</tr>
<tr class="odd">
<td>lme4qtl</td>
<td><code>relmatLmer(myTrait ~ myCovariate + (1|myID), myData, relmat = list(myID = myMatrix))</code></td>
</tr>
</tbody>
</table>

<table style="width:46%;">
<colgroup>
<col width="15%" />
<col width="30%" />
</colgroup>
<thead>
<tr class="header">
<th>Package</th>
<th>Binary response</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>stats</td>
<td><code>glm(myStatus ~ 1, myData, family = binomial)</code></td>
</tr>
<tr class="even">
<td>lme4</td>
<td><code>glmer(myStatus ~ (1|myID), myData, family = binomial)</code></td>
</tr>
<tr class="odd">
<td>lme4qtl</td>
<td><code>relmatGlmer(myStatus ~ (1|myID), myData, relmat = list(myID = myMatrix), family = binomial)</code></td>
</tr>
</tbody>
</table>

Note that rownames/colnames of `myMatrix` have to be values of `myID` variable, so matching between relationship matrix and grouping variable is possible. The order [doesn't matter](https://github.com/variani/lme4qtl/issues/3#issuecomment-346905978).

Installation
------------

You can install the development version from [GitHub](https://github.com/variani/lme4qtl) with:

``` r
# install.packages("devtools")
devtools::install_github("variani/lme4qtl")
```

The official release on [CRAN](https://CRAN.R-project.org) is [pending](https://github.com/variani/lme4qtl/issues/9).

Citation
--------

To cite the `lme4qtl` package in publications use:

      Ziyatdinov et al., lme4qtl: linear mixed models with flexible
      covariance structure for genetic studies of related individuals, 
      BMC Bioinformatics (2018)

Contact
-------

You are welcome to submit suggestions and bug-reports at <https://github.com/variani/lme4qtl/issues>.

Example
-------

``` r
library(lme4)
library(lattice)
```

``` r
library(lme4qtl)
```

``` r
packageVersion("lme4qtl")
#> [1] '0.2.1'
```

Load simulated data, phenotypes `dat40` and the kinship matrix `kin2`.

``` r
data(dat40, package = "lme4qtl")

dim(dat40)
#> [1] 234  10
dim(kin2)
#> [1] 234 234

head(dat40)
#>     ID  trait1  trait2 AGE FAMID  FA  MO SEX trait1bin trait2bin
#> 7  101 8.41954 9.67925  50    10   0   0   1         0         0
#> 14 102 5.47141 4.31886  44    10   0   0   2         0         0
#> 21 103 9.66547 7.00735  34    10 101 102   2         0         0
#> 28 104 6.27092 8.59257  41    10 101 102   1         0         0
#> 35 105 7.96814 7.60801  36    10 101 102   1         0         0
#> 42 106 8.29865 8.17634  37    10 101 102   2         0         0
kin2[1:5, 1:5] # nuclear family with 2 parents and 3 kids
#> 5 x 5 sparse Matrix of class "dsCMatrix"
#>     11  12  13  14  15
#> 11 1.0 .   0.5 0.5 0.5
#> 12 .   1.0 0.5 0.5 0.5
#> 13 0.5 0.5 1.0 0.5 0.5
#> 14 0.5 0.5 0.5 1.0 0.5
#> 15 0.5 0.5 0.5 0.5 1.0
```

Fit a model for continuous trait with two random effects, family-grouping `(1|FAM)` and additive genetic `(1|ID)`.

``` r
m1 <- relmatLmer(trait1 ~ AGE + SEX + (1|FAMID) + (1|ID), dat40, relmat = list(ID = kin2))
#> boundary (singular) fit: see ?isSingular
#> boundary (singular) fit: see ?isSingular
m1
#> Linear mixed model fit by REML ['lmerMod']
#> Formula: trait1 ~ AGE + SEX + (1 | FAMID) + (1 | ID)
#>    Data: dat40
#> REML criterion at convergence: 963.3853
#> Random effects:
#>  Groups   Name        Std.Dev.
#>  ID       (Intercept) 2.2988  
#>  FAMID    (Intercept) 0.0000  
#>  Residual             0.7856  
#> Number of obs: 224, groups:  ID, 224; FAMID, 39
#> Fixed Effects:
#> (Intercept)          AGE         SEX2  
#>    7.563248     0.008314    -0.364197  
#> convergence code 0; 1 optimizer warnings; 0 lme4 warnings
```

Get a point estimate of heritability (h2), the proportion of variance explained by `(1|ID)`.

``` r
lme4::VarCorr(m1)
#>  Groups   Name        Std.Dev.
#>  ID       (Intercept) 2.29880 
#>  FAMID    (Intercept) 0.00000 
#>  Residual             0.78562
lme4qtl::VarProp(m1)
#>        grp        var1 var2      vcov     sdcor      prop
#> 1       ID (Intercept) <NA> 5.2845002 2.2988041 0.8954191
#> 2    FAMID (Intercept) <NA> 0.0000000 0.0000000 0.0000000
#> 3 Residual        <NA> <NA> 0.6172059 0.7856245 0.1045809
```

Profile the variance components (h2) to get the 95% confidence intervals. The method functions `profile` and `confint` are implemented in lme4. Note that a different model `m2` is used, because profiling is prone to errors/warnings if model fit is poor.

``` r
m2 <- relmatLmer(trait2 ~ (1|ID), dat40, relmat = list(ID = kin2)) 
VarProp(m2)
#>        grp        var1 var2     vcov    sdcor      prop
#> 1       ID (Intercept) <NA> 5.573272 2.360778 0.7723589
#> 2 Residual        <NA> <NA> 1.642638 1.281654 0.2276411

prof <- profile(m2)
#> Warning in zetafun(np, ns): NAs detected in profiling
prof_prop <- lme4qtl::varpropProf(prof) # convert to proportions
confint(prof_prop)
#>                 2.5 %    97.5 %
#> .sigprop01  0.5292158 0.9157910
#> .sigmaprop  0.0655726 0.4652175
#> (Intercept) 7.2745157 8.4237085
```

``` r
densityplot(prof)
```

<img src="man/figures/README-plot_prof-1.png" width="50%" />

``` r
densityplot(prof_prop)
```

<img src="man/figures/README-plot_prof-2.png" width="50%" />

``` r
try(splom(prof)) 
#> Error in if (singfit) warning("splom is unreliable for singular fits") : 
#>   missing value where TRUE/FALSE needed

prof_clean <- na.omit(prof) # caution: NAs are indicators of poor fits
splom(prof_clean)
```

<img src="man/figures/README-plot_prof_splom-1.png" width="50%" />

Fit a model with genetic and residual variances that differ by gender (sex-specificity model). The formula syntax with `dummy` (see `?lme4::dummy`) is applied to the residual variance `(1|RID)` to cancel the residual correlation.

``` r
dat40 <- within(dat40, RID <- ID) # replicate ID 

m4 <- relmatLmer(trait2 ~ SEX + (0 + SEX|ID) + (0 + dummy(SEX)|RID), dat40, relmat = list(ID = kin2)) 
VarCorr(m4)
#>  Groups   Name       Std.Dev.   Corr 
#>  ID       SEX1       1.94400138      
#>           SEX2       2.64404940 0.826
#>  RID      dummy(SEX) 0.00050224      
#>  Residual            1.22780606
```

An example of parameter constraints that make the genetic variance between genders equal.

``` r
m4_vareq <- relmatLmer(trait2 ~ SEX + (0 + SEX|ID) + (0 + dummy(SEX)|RID), dat40, relmat = list(ID = kin2),
  vcControl = list(vareq = list(id = c(1, 2, 3)))) 
VarCorr(m4_vareq)
#>  Groups   Name       Std.Dev. Corr 
#>  ID       SEX1       2.47777       
#>           SEX2       2.47777  0.746
#>  RID      dummy(SEX) 0.95827       
#>  Residual            0.72728
```

Another example of parameter constraint that implies the genetic correlation between genders equal to 1.

``` r
m4_rhog1 <- relmatLmer(trait2 ~ SEX + (0 + SEX|ID) + (0 + dummy(SEX)|RID), dat40, relmat = list(ID = kin2),
  vcControl = list(rho1 = list(id = 3))) 
VarCorr(m4_rhog1)
#>  Groups   Name       Std.Dev.  Corr 
#>  ID       SEX1       1.7627785      
#>           SEX2       2.5782330 1.000
#>  RID      dummy(SEX) 0.0014823      
#>  Residual            1.4147793
```

Fit a model for binary trait.

``` r
m3 <- relmatGlmer(trait1bin ~ (1|ID), dat40, relmat = list(ID = kin2), family = binomial(probit))
m3
#> Generalized linear mixed model fit by maximum likelihood (Laplace
#>   Approximation) [glmerMod]
#>  Family: binomial  ( probit )
#> Formula: trait1bin ~ (1 | ID)
#>    Data: dat40
#>       AIC       BIC    logLik  deviance  df.resid 
#>  218.1325  225.0432 -107.0663  214.1325       232 
#> Random effects:
#>  Groups Name        Std.Dev.
#>  ID     (Intercept) 0.7669  
#> Number of obs: 234, groups:  ID, 234
#> Fixed Effects:
#> (Intercept)  
#>      -1.242
```


NEWS
lme4qtl 0.2.2
=============

- Throws a warning when misspecified IDs in `relamt` argument. 
  Related to issue #20 and SO question 
  https://stats.stackexchange.com/questions/417160/linear-mixed-effects-model-i-cant-seem-to-avoid-either-convergence-errors-or
    
lme4qtl 0.2.1
=============

- Merged with #18 and its extension #19.
  The new update of Z-matrices by the Khatri-Rao product (proposed by @deruncie)
  is more elegant and in accordance with the lme4 implementation;
  see https://github.com/variani/lme4qtl/pull/18#issuecomment-510646099.
  Now creating models with polygenic slopes is possible;
  see code examples (model `m6` and `m6b`) in #18.

lme4qtl 0.1.11
=============

- Fixed issue #1 (https://github.com/variani/lme4qtl/issues/1).
  lme4 threw the "error with reference class field ‘Ut’, should be from class “dgCMatrix”",
  when lme4qtl decomposed the custom covariance matrix as A = R'R.
  That was solved using `as(R, "dgCMatrix")` instead of `as.Matrix(R) in `relfac` functions.

lme4qtl 0.1.10
=============

- Fixed issues:
    - #5 (https://github.com/variani/lme4qtl/issues/5): repeated measurements and `relmatGlmer`

lme4qtl 0.1.9
=============

- The first public version hosted at https://github.com/variani/lme4qtl

lme4qtl 0.1.8
=============

- The SVD method is replaced by EVD in `relfac` fucntion.
  Now the decompoistion K = LL' works for rank-deficient matrices such as (1|FAM).
  The test file `test-relfac.R` is updated with necessary checks.

lme4qtl 0.1.7
=============

* Towards the official release before submitting the article to Bionf.

lme4qtl 0.1.6
=============

* The use of variables given in `relmat` are revised.
  Now the cases when when these variables are characters or integers
  seems to work OK.
  The point is that in the order of data rows (used when passed from formula to matrices).
  The example that may fail without this update is http://kbroman.org/lmmlite/assets/lmmlite.html
  (the heritability is estimated as `0`).

lme4qtl 0.1.5
=============

* The release prepared for the internal use of collaborators.
  Two main functions were tested: `relmatLmer` and `relmatGlmer`.

lme4qtl 0.1.4
=============

* The package was renamed from `solaris` to `lme4qtl`.
* The old code was depreciated and the old R files are moved to `docs/R/old`.
* The package was hadleversed as much as needed to pass `devtools::check()`.

solaris 0.1.3
==================

* Updated `solaris` function to hadle `relmat` parameter better.
  Now `step` funtion from `lmerTest` package works.

solaris 0.1.2
==================

* Added `relranef` function. This function extracts the random effect vectors `u`
  when relative matrices were introduced for the random effects.
  See example code in `ins/examples/function-realranef.R`
  
solaris 0.1.1
==================

* Added `stepRaef` function. This function takes all random effect in a model
  and then follow `step`-like procedure to filter out those significant effect.
  Likelihood ratio test (LRT) is used.


R/relmatLmer.R
#---------------------
# Function `relmatLmer`
#---------------------

#' Fit a linear mixed model (LMM) with relative matrices.
#'
#' Its implementation follows that of lmer function in lme4 R package.
#'
#' @export
relmatLmer <- function(...)
{
  mc <- match.call()
  
  env <- parent.frame(1)
  env$relmat_lmer <- relmat_lmer
  
  mc[[1]] <- quote(relmat_lmer)
  eval(mc, env)
}

relmat_lmer <- function(formula, data = NULL, REML = TRUE,
  control = lmerControl(), start = NULL,
  verbose = 0L, subset, weights, na.action, offset,
  contrasts = NULL, devFunOnly = FALSE, 
  ...,
  # relLmer-specific argument in comparison with lmer function
  method.relfac = "auto",
  debug = 0,
  modOnly = FALSE, controlOnly = FALSE,
  vcControl = list(),
  relmat = list(), check.nobs = "ignore", calc.derivs = TRUE 
)
{
  mc <- mcout <- match.call()
  missCtrl <- missing(control)
  
  ## see functions in modular.R for the body ...
  if(!missCtrl && !inherits(control, "lmerControl")) {
    if(!is.list(control)) {
      stop("'control' is not a list; use lmerControl()")
    }
  
    ## back-compatibility kluge
    warning("passing control as list is deprecated: please use lmerControl() instead", immediate. = TRUE)
    control <- do.call(lmerControl, control)
  }
  
  #----
  # relmatLmer-specific part
  #----
  #relnms <- names(relmat)
  #stopifnot(all(relnms %in% names(data)))
  #relcls <- laply(relnms, function(x) class(data[[x]]))
  #if(!all(relcls == "factor")) {
  #  stop("all variables related to 'relmat' to be factors")
  #}
  
  if(check.nobs == "ignore") {
    control$checkControl$check.nobs.vs.rankZ <- "ignore"
    control$checkControl$check.nobs.vs.nlev <- "ignore"
    control$checkControl$check.nobs.vs.nRE <- "ignore"
  }
  control$calc.derivs <- calc.derivs
  
  if (!is.null(list(...)[["family"]])) {
    stop("calling lmer with 'family' is deprecated; please use glmer() instead")
  }
  
  if(controlOnly) return(control)
  
  mc$control <- control ## update for back-compatibility kluge
  ## https://github.com/lme4/lme4/issues/50

  ## parse relmat, etc arguments
  nrelmat <- length(relmat)
  
  ## parse data and formula
  mc[[1]] <- quote(lme4::lFormula)
  #lmod <- eval(mc, parent.frame(1L)) ## parse data and formula
  
  #----
  # relmatLmer-specific part
  #----
  # remove call arguments which are specific to `relmat_lmer` and to be removed before calling lme4::lFormula
  args <- c("method.relfac", "modOnly", "controlOnly", "debug", "vcControl", 
    "relmat", "check.nobs", "calc.derivs")
  for(arg in args) {
    if(arg %in% names(mc)) {
      mc[[arg]] <- NULL
    }
  }

  # @ https://github.com/lme4/lme4/commit/21dfb0114562a36c59ab6b90e4dde7311c7e4c70
  #lmod <- eval(mc, environment(formula)) ## parse data and formula
  lmod <- eval(mc, parent.frame(1L))
  
  #-------------------------------
  # start of solaris-specific code
  #-------------------------------
  stopifnot(is.list(relmat), length(names(relmat)) == length(relmat))
  relfac <- list()
  
  # list of random term factor
  flist <- lmod$reTrms[["flist"]]
  fnmns <- names(flist) # e.g. (1|gr) + (1|id) 
  
  # list of random term factors with custom covariances
  # - specified in `relmat` argument
  # - need special processing
  relnms <- names(relmat)

  if(!all(relnms %in% fnmns)) {
    warning("not all relmat ID variables match RE ID terms")
  }

  for(i in seq_along(fnmns)) {
    fn <- fnmns[i]
    
    if(fn %in% relnms) { # update only when matching; otherwise, ignore
      if(debug) {
        cat(" - Zlist element", fn, "\n")
      }
      
      # check names
      zn <- lmod$fr[, fn]
      if(class(zn) != "factor") { # convert to factor
        zn <- as.factor(zn) 
      }
      zn.unique <- levels(zn)

      stopifnot(!is.null(rownames(relmat[[fn]])))
      rn <- rownames(relmat[[fn]])

      stopifnot(all(zn.unique %in% rn))
      
      # compute a relative factor R: K = R'R
      # See lme4qtl:::relfac
      K <- Matrix::Matrix(relmat[[fn]][zn.unique, zn.unique], sparse = TRUE)
      R <- relfac(K, method.relfac)
      relfac[[fn]] <- R

      # compute a substitution Z*
      # 1. for random effects ~(1|f): Z* = R Z'
      # 2. for random effects ~(r | f)
      #   r is used to construct the "raw random effects model matrix Xi (n x pi) for term i
      #   f is used to for the grouping factor model matrix Ji (n x qi) for a factor with qi levels.
      #       These levels correspond to levels(flist[[fnmn]]) and are used to index the rows/columns of Ki
      # Following Bates, et al "Fitting Linear Mixed-Effects Models Using lme4",
      # these are combined into the matrix Zi = t(KhatriRao(t(Ji), t(Xi)))
      # 
      # We want to modify Zi to: t(KhatriRao(t(Ji %*% t(R)), t(Xi))), 
      # where t(R) %*% R = K, eg. Cholesky
      #
      # This is equal to Zi %*% kronecker(t(K_sqrt),diag(1,pi))
      #
      # See github.com/variani/lme4qtl/pull/18 contributed by github.com/deruncie
      
      pi <- length(lmod$reTrms$cnms[[i]])
      Zi_t <- lmod$reTrms$Ztlist[[i]] 
      Zi_t <- kronecker(R, diag(1, pi)) %*% Zi_t # t(Z*)

      # put the new t(Z*) back into the appropriate slot `Ztlist`
      lmod$reTrms$Ztlist[[i]] <- Zi_t
    }
  }
  # update the full Zt matrix (the slot `Zt`) by combining all Zt matrices (the slot `Ztlist`)
  lmod$reTrms[["Zt"]] <- do.call(rBind, lmod$reTrms$Ztlist)
  #-------------------------------
  # end of relmatLmer-specific code
  #------------------------------- 
    
  mcout$formula <- lmod$formula
  lmod$formula <- NULL
  
  if(modOnly) return(lmod)
  
  ## create deviance function for covariance parameters (theta)
  devfun <- do.call(mkLmerDevfun, c(lmod,
    list(start = start, verbose = verbose, control = control)))
  # print(with(environment(devfun), pp$theta))
  
  #----
  # vcControl-specific update
  #----  
  if(length(vcControl)) {
    # local update function  
    update_theta_vcControl <- function(theta, vcControl)
    {
      nth <- length(theta)
    
      # vcControl.vareq 
      # Example: c(1, 2, 3)
      vcControl.vareq <- vcControl[["vareq"]]  
      if(length(vcControl.vareq)) {
        for(i in 1:length(vcControl.vareq)) {
          ind <- vcControl.vareq[[i]]
          stopifnot(length(ind) == 3)
          stopifnot(all(ind <= nth))
          
          # t1^2 = t12^2  + t2^2
          theta[ind[1]] <- sqrt(theta[ind[2]]*theta[ind[2]] + theta[ind[3]]*theta[ind[3]])
        }
      }
      
      # vcContol.rho1
      # Example: 3
      vcControl.rho1 <- vcControl[["rho1"]]
      if(length(vcControl.rho1)) {
        for(i in 1:length(vcControl.rho1)) {
          ind <- vcControl.rho1[[i]]
          stopifnot(length(ind) == 1)
          stopifnot(all(ind <= nth))
          
          # t2 = 0          
          theta[ind[1]] <- 0
        }
      }

      # vcContol.rho0
      # Example: 2      
      vcControl.rho0 <- vcControl[["rho0"]]
      if(length(vcControl.rho0)) {
        for(i in 1:length(vcControl.rho0)) {
          ind <- vcControl.rho0[[i]]
          stopifnot(length(ind) == 1)
          stopifnot(all(ind <= nth))
          
          # t12 = 0   
          theta[ind[1]] <- 0
        }
      }
      
      # vcContol.rho
      # Example: 2, 3      
      vcControl.rho <- vcControl[["rho"]]
      if(length(vcControl.rho)) {
        for(i in 1:length(vcControl.rho)) {
          ind <- vcControl.rho[[i]][[1]]

          stopifnot(length(ind) == 2)
          stopifnot(all(ind <= nth))
          
          val <- vcControl.rho[[i]][[2]]
          stopifnot(val != 0)
          stopifnot(val > -1)
          stopifnot(val < 1)
          
          k <- val / sqrt(1 - val*val)

          # t12 = rho / sqrt(1 - rho^2) t2   
          theta[ind[1]] <- k * theta[ind[2]]
        }
      }
      
      # vcContol.th1
      # Example: 1      
      vcControl.th1 <- vcControl[["th1"]]
      if(length(vcControl.th1)) {
        for(i in 1:length(vcControl.th1)) {
          ind <- vcControl.th1[[i]]
          stopifnot(length(ind) == 1)
          stopifnot(all(ind <= nth))
          
          # t1 = 1   
          theta[ind[1]] <- 1
        }
      }

      # vcContol.th0
      # Example: 0    
      vcControl.th0 <- vcControl[["th0"]]
      if(length(vcControl.th0)) {
        for(i in 1:length(vcControl.th0)) {
          ind <- vcControl.th0[[i]]
          stopifnot(length(ind) == 1)
          stopifnot(all(ind <= nth))
          
          # t1 = 0   
          theta[ind[1]] <- 0
        }
      }            

      # vcControl.hom3 
      # Example: c(1, 2, 3, 4, 5, 6)
      vcControl.hom3 <- vcControl[["hom3"]]  
      if(length(vcControl.hom3)) {
        for(i in 1:length(vcControl.hom3)) {
          ind <- vcControl.hom3[[i]]
          stopifnot(length(ind) == 6)
          stopifnot(all(ind <= nth))
          
          # t3 = t2
          theta[ind[3]] <- theta[ind[2]]
          # t4 = sqrt(t1^2 - t2^2)
          stopifnot(abs(theta[ind[1]]) >= abs(theta[ind[2]]))
          theta[ind[4]] <- sqrt(theta[ind[1]]*theta[ind[1]] - 
            theta[ind[2]]*theta[ind[2]])
          # t5 = (t1*t2 - t2^2) / t4
          theta[ind[5]] <- (theta[ind[1]]*theta[ind[2]] - 
            theta[ind[2]]*theta[ind[2]]) / theta[ind[4]]
          # t6 = sqrt(t4^2 - t5^2)
          stopifnot(abs(theta[ind[4]]) >= abs(theta[ind[5]]))
          theta[ind[6]] <- sqrt(theta[ind[4]]*theta[ind[4]] - 
            theta[ind[5]]*theta[ind[5]])
        }
      }
   
         
      return(theta)
    }

    devfun.wrapper <- function(theta) {
      th <- theta
      th <- update_theta_vcControl(th, vcControl)
      
      devfun(th)
    }
    environment(devfun.wrapper) <- environment(devfun)
    assign("devfun", devfun, envir = environment(devfun.wrapper))
    assign("vcControl", vcControl, envir = environment(devfun.wrapper))
    assign("update_theta_vcControl", update_theta_vcControl, envir = environment(devfun.wrapper))
  }
  #----
  # end of vcControl-specific update
  #----  
  
  if(devFunOnly) return(devfun)
  
  ## optimize deviance function over covariance parameters
  if (control$optimizer=="none") {
    #----
    # relmatLmer-specific update
    #----
    if(is.null(start)) {
      opt <- list(par = NA, fval = NA, conv = 1000, message = "no optimization")
    }
    else if("theta" %in% names(start)) {
      opt <- list(par = as.numeric(start$theta), fval = NA, conv = 1000, message = "no optimization + start")
      if(!is.null(start)) {
        if("theta" %in% names(start)) {
          theta <- start$theta
          #devfun(pp$theta <- theta)
          #assign("pp$theta", theta)
          devfun(theta)
        }
      }
    }
  } else {
    if(length(vcControl)) {
      opt <- optimizeLmer(devfun.wrapper,
        optimizer=control$optimizer,
        restart_edge=control$restart_edge,
        boundary.tol=control$boundary.tol,
        control=control$optCtrl,
        verbose=verbose,
        start=start,
        calc.derivs=control$calc.derivs,
        use.last.params=control$use.last.params)    
    } else {
      opt <- optimizeLmer(devfun,
        optimizer=control$optimizer,
        restart_edge=control$restart_edge,
        boundary.tol=control$boundary.tol,
        control=control$optCtrl,
        verbose=verbose,
        start=start,
        calc.derivs=control$calc.derivs,
        use.last.params=control$use.last.params)
        
      cc <- lme4:::checkConv(attr(opt,"derivs"), opt$par,
        ctrl = control$checkConv,
        lbound = environment(devfun)$lower)
    }
  }

  if(length(vcControl)) {
    th <- opt$par
    th <- update_theta_vcControl(th, vcControl)
    devfun(opt$par <- th)

    #mkMerMod(environment(devfun), opt.eq, lmod$reTrms, fr = lmod$fr)
    
    opt <- list(par = th, feval = opt$feval, fval = NA, conv = opt$convergence, message = "no optimization + par from devfun.wrapper")
    mod <- mkMerMod(environment(devfun), opt, lmod$reTrms, fr = lmod$fr, mcout) ## prepare output
  } else {
    cc <- lme4:::checkConv(attr(opt,"derivs"), opt$par,
      ctrl = control$checkConv,
      lbound=environment(devfun)$lower)
  
    mod <- mkMerMod(environment(devfun), opt, lmod$reTrms, fr = lmod$fr, mcout, lme4conv = cc) ## prepare output
  }
  
  # adhoc solution to store `relmatLmer` variables in an object of `lmerMod` class 
  mod@optinfo$relmat <- list(relfac = relfac)
  
  return(mod)
}

relmatLmer_naive <- function(formula, data = NULL, 
  start = NULL,
  relmat = list()
)
{
  # formula
  control <- lmerControl(check.nobs.vs.rankZ = "ignore", 
    check.nobs.vs.nlev = "ignore", check.nobs.vs.nRE = "ignore")
  
  # lmod
  lmod <- lFormula(formula, data, control = control)
  
  #-------------------------------
  # start of relmatLmer-specific code
  #-------------------------------
  stopifnot(is.list(relmat), length(names(relmat)) == length(relmat))
  relnms <- names(relmat)
  relfac <- relmat
  flist <- lmod$reTrms[["flist"]]   ## list of factors

  ind <- (relnms %in% names(flist))
  if(any(ind)) {
    ## random-effects design matrix components
    Ztlist <- lmod$reTrms[["Ztlist"]]
    
    asgn <- attr(flist, "assign")
    for(i in seq_along(relnms[ind])) {
      
      relmati <- relnms[ind][i]
      if(!(relmati %in% names(flist))) {
        stop("a relationship matrix must be (", relmati, ")",
          " associated with only one random effects term (", paste(names(flist), collapse = ", "), ")")
      }
      tn <- which(relmati == names(flist))
      fn <- names(flist)[tn]
      
      zn <- lmod$fr[, fn]
      if(class(rownames(relmat[[i]])) == "character") {
        zn <- as.character(zn)
      }
      
      relmat[[i]] <- Matrix::Matrix(relmat[[i]][zn,zn], sparse = TRUE)
      stopifnot(nrow(relmat[[i]]) == nrow(lmod$fr))
    
      relfac[[i]] <- Matrix::chol(relmat[[i]])
      Ztlist[[i]] <-  relfac[[i]] %*% Ztlist[[i]]
    }
    lmod$reTrms[["Ztlist"]] <- Ztlist
  }
  lmod$reTrms[["Zt"]] <- do.call(rBind, Ztlist)
  #-------------------------------
  # end of relmatLmer-specific code
  #-------------------------------
  
  devfun <- do.call(mkLmerDevfun, c(lmod,
    list(start = start)))#, verbose = verbose, control = control)))
  devfun(opt$par <- start)
  
  opt <- list(par = as.numeric(start$theta), fval = NA,conv = 1000, message="start copied")
  mkMerMod(environment(devfun), opt, lmod$reTrms, fr = lmod$fr)
}
